c2rust-analyze/src/annotate.rs

```diff
@@ -0,0 +1,62 @@
+use rustc_middle::ty::TyCtxt;
+use rustc_span::{FileName, Span};
+use std::collections::HashMap;
+use std::fmt::Display;
+
+pub struct AnnotationBuffer<'tcx> {
+    tcx: TyCtxt<'tcx>,
+    /// Map from `file_idx` to a list of annotations as `(line_number, text)` pairs.
+    m: HashMap<usize, Vec<(usize, String)>>,
+}
+
+impl<'tcx> AnnotationBuffer<'tcx> {
+    pub fn new(tcx: TyCtxt<'tcx>) -> AnnotationBuffer<'tcx> {
+        AnnotationBuffer {
+            tcx,
+            m: HashMap::new(),
+        }
+    }
+
+    pub fn _clear(&mut self) {
+        self.m.clear();
+    }
+
+    pub fn emit(&mut self, span: Span, msg: impl Display) {
+        let sm = self.tcx.sess.source_map();
+
+        let span = span.source_callsite();
+        let pos = span.lo();
+        let file_idx = sm.lookup_source_file_idx(pos);
+        let sf = &sm.files()[file_idx];
+        let line = sf.lookup_line(pos).unwrap_or(0);
+
+        let src = sm
+            .span_to_snippet(span)
+            .unwrap_or_else(|_| "<error>".into());
+        let src = src.split_ascii_whitespace().collect::<Vec<_>>().join(" ");
+        let (src1, src2, src3) = if src.len() > 20 {
+            (&src[..15], " ... ", &src[src.len() - 5..])
+        } else {
+            (&src[..], "", "")
+        };
+        self.m.entry(file_idx).or_insert_with(Vec::new).push((
```

Comment:
`or_default()` is shorter than `or_insert_with(Vec::new)`.

Reply:
I prefer the `or_insert_with(Vec::new)` version as a hint to the reader that the map values are `Vec`s.